README.md
# Excalidraw VSCode Plugin

The [Excalidraw](https://excalidraw.com/) plugin in VSCode

## install

[vscode-excalidraw-plugin](https://marketplace.visualstudio.com/items?itemName=jkchao.vscode-excalidraw-plugin)

## Features

- Edit `.excalidraw`, `.excalidraw.json`, `.exd.json` file in editor.
- Create `.excalidraw`, `.excalidraw.json`, `.exd.json` file, and open it.

## Extension Settings

| Setting        | type    | default      | Required | desc                                                          |
| -------------- | ------- | ------------ | -------- | ------------------------------------------------------------- |
| excalidraw.defaultBackgroundColor   | string  | "#ffffff"   | false     | The Default Background Color             |
## DEMO

### 1.Base

![demo1](./images/demo01.png)

### 2.Export to PNG/SVG

![demo2](./images/demo02.gif)

## TODO

https://stackoverflow.com/questions/62102767/how-to-use-window-crypto-subtle-in-vscode-webview


CHANGELOG.md
# Change Log

## 0.1.5

-   Update Excalidraw

## 0.0.7

-   Update Excalidraw

## 0.0.6

-   Support `.excalidraw`, `.excalidraw.json`, `.exd.json` file.

## 0.0.5

-   Remove `Reset the canvas` Btn.
-   Add config, the defaultBackgroundColor
-   Update Excalidraw

## 0.0.4

-   Use the last Excalidraw
-   Update logo

## 0.0.3

-   fix bug [#6](https://github.com/jkchao/vscode-excalidraw/issues/6)
-   fix bug [#7](https://github.com/jkchao/vscode-excalidraw/issues/7)
-   update demo

## 0.0.2

-   update demo

## 0.0.1

-   Initial release


src/webviewContent.ts
import * as vscode from 'vscode';
import { join } from 'path';

export class WebViewContent {
  private buildPath: string;
  constructor(rootPath: string) {
    this.buildPath = join(rootPath, 'build');
  }

  private readFileOnDisk(path: string) {
    return vscode.Uri.file(join(this.buildPath, path)).with({
      scheme: 'vscode-resource'
    });
  }

  public createWebViewContent() {
    // script
    const manifest = require(`${this.buildPath}/asset-manifest.json`);
    console.log(manifest);
    const mainScript = manifest.files['main.js'];
    const mainStyle = manifest.files['main.css'];
    const runtimeScript = manifest.files['runtime-main.js'];
    // chunk
    const chunkScript = [];
    for (const key in manifest.files) {
      if (key.endsWith('.chunk.js') && manifest.files.hasOwnProperty(key)) {
        // finding their paths on the disk
        const chunk = this.readFileOnDisk(manifest.files[key]);
        chunkScript.push(chunk);
      }
    }

    const runtimeScriptOnDisk = this.readFileOnDisk(runtimeScript);
    const mainScriptOnDisk = this.readFileOnDisk(mainScript);
    const styleScriptOnDisk = this.readFileOnDisk(mainStyle);

    return `<!DOCTYPE html>
    <html lang="en">
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>ExcaliDraw</title>

        <link rel="stylesheet" type="text/css" href="${styleScriptOnDisk}">
        <link rel="stylesheet" href="fonts.css" />
        <link rel="preload" href="${this.buildPath}/FG_Virgil.woff2" as="font" type="font/woff2" crossorigin="anonymous" />
        <link rel="preload" href="${this.buildPath}/Cascadia.woff2" as="font" type="font/woff2" crossorigin="anonymous" />

        <style>
        
        .LoadingMessage {
          position: fixed;
          top: 0;
          right: 0;
          bottom: 0;
          left: 0;
          z-index: 999;
          display: flex;
          align-items: center;
          justify-content: center;
          pointer-events: none
      }

      .LoadingMessage span {
          background-color: rgba(255, 255, 255, .8);
          border-radius: 5px;
          padding: .8em 1.2em;
          font-size: 1.3em
      }

      .visually-hidden {
          position: absolute !important;
          height: 1px;
          width: 1px;
          overflow: hidden;
          clip: rect(1px 1px 1px 1px);
          clip: rect(1px, 1px, 1px, 1px);
          white-space: nowrap
      }
        </style>
        
    </head>
    <body>
      <div id="root">
          <div class="LoadingMessage"><span>Loading scene...</span></div>
      </div>
      <script>
        // https://github.com/Microsoft/vscode/issues/48464
        Object.defineProperty(document, 'cookie', {
          get: () => '',
          set: () => ''
        });
        Object.defineProperty(document, 'localStorage', {
          get: () => '',
          set: () => ''
        });
      </script>
      <script src="${runtimeScriptOnDisk}"></script>
      ${chunkScript.map((item: vscode.Uri) => `<script src="${item}"></script>`)}
      <script src="${mainScriptOnDisk}"></script>
      <script>
        window.error = e => {
          console.log(e);
        }
      </script>
    </body>
    </html>`;
  }
}


src/excalidrawEditor.ts
import * as vscode from 'vscode';
import { join } from 'path';
import html from './webviewHtml.html';
import { EXCALIDRAW_DEFAULT_BACKGROUNDCOLOR } from './constants';

// https://github.com/webpack/webpack/issues/4175
declare const __non_webpack_require__: any;

export class ExcalidrawEditorProvider implements vscode.CustomExecution {
    public static register(context: vscode.ExtensionContext): vscode.Disposable {
        const provider = new ExcalidrawEditorProvider(context);
        const providerRegistration = vscode.window.registerCustomEditorProvider(ExcalidrawEditorProvider.viewType, provider, { webviewOptions: { retainContextWhenHidden: true } });
        return providerRegistration;
    }

    private static readonly viewType = 'excalidraw.openEditor';
    private buildPath: string;
    private elements = [];
    private config = vscode.workspace.getConfiguration('excalidraw');
    private appState = {};

    constructor(private context: vscode.ExtensionContext) {
        this.buildPath = join(context.extensionPath, 'excalidraw', 'build');
    }

    private readFileOnDisk(path: string) {
        return vscode.Uri.file(join(this.buildPath, path)).with({
            scheme: 'vscode-resource',
        });
    }
    private async updateExcalidrawContent(document: vscode.TextDocument, data: string, key: 'updateExcalidraw' | 'updateExcalidrawState') {
        // https://github.com/excalidraw/excalidraw/blob/c427aa3cce801bef4dd9107e1044d3a4f61a201e/src/data/json.ts#L12
        const parseData = JSON.parse(data);

        if (key === 'updateExcalidraw') {
            this.elements = parseData;
        }

        if (key === 'updateExcalidrawState') {
            this.appState = parseData;
        }

        // TODO AppState, viewBackgroundColor
        const result = JSON.stringify({
            type: 'excalidraw',
            elements: this.elements.filter((element: any) => !element.isDeleted),
            appState: this.appState,
        });

        if (result === document.getText()) {
            return false;
        }

        const workspaceEdit = new vscode.WorkspaceEdit();

        workspaceEdit.replace(document.uri, new vscode.Range(0, 0, document.lineCount, 0), result);

        try {
            await vscode.workspace.applyEdit(workspaceEdit);
        } catch (error) {
            // ...
        }

        // const buf = Buffer.from(result);
        // vscode.workspace.fs.writeFile(document.uri, buf);
    }

    private createWebViewContent(webview: vscode.Webview) {
        // script
        // https://github.com/webpack/webpack/issues/4175
        const manifest = __non_webpack_require__(join(this.buildPath, 'asset-manifest.json'));
        const mainScript = manifest.files['main.js'];
        const mainStyle = manifest.files['main.css'];
        const runtimeScript = manifest.files['runtime-main.js'];

        // chunk
        const chunkScript = [];
        for (const key in manifest.files) {
            if (key.endsWith('.chunk.js') && manifest.files.hasOwnProperty(key)) {
                const chunk = this.readFileOnDisk(manifest.files[key]);
                chunkScript.push(chunk);
            }
        }

        const runtimeScriptOnDisk = this.readFileOnDisk(runtimeScript);

        const mainScriptOnDisk = this.readFileOnDisk(mainScript);
        const styleScriptOnDisk = this.readFileOnDisk(mainStyle);
        const virgilFontOnDisk = this.readFileOnDisk('FG_Virgil.woff2');
        const cascadiaFontOnDisk = this.readFileOnDisk('Cascadia.woff2');
        const fontPathOnDisk = this.readFileOnDisk('fonts.css');

        // const patchedHtml = html
        //     // .replace(new RegExp('{{extensionPath}}', 'g'), this.buildPath)
        //     .replace(new RegExp('{{fontPath}}', 'g'), fontPathOnDisk.toString())
        //     .replace(new RegExp('{{virgilFont}}', 'g'), virgilFontOnDisk.toString())
        //     .replace(new RegExp('{{cascadiaFont}}', 'g'), cascadiaFontOnDisk.toString())
        //     .replace(new RegExp('{{styleScriptOnDisk}}', 'g'), styleScriptOnDisk.toString())
        //     .replace(new RegExp('{{chunkScript}}', 'g'), chunkScript.map(item => `${item.scheme}:/${item.path}`).join(','))
        //     .replace(new RegExp('{{runtimeScriptOnDisk}}', 'g'), runtimeScriptOnDisk.toString())
        //     .replace(new RegExp('{{mainScriptOnDisk}}', 'g'), mainScriptOnDisk.toString());

        // return patchedHtml;

        return `
            <!DOCTYPE html>
            <html lang="en">
            
            <head>
                <meta charset="UTF-8" />
                <meta name="viewport" content="width=device-width, initial-scale=1.0" />
                <title>ExcaliDraw</title>
            

                <style>
                    /* http://www.eaglefonts.com/fg-virgil-ttf-131249.htm */
                    @font-face {
                        font-family: 'Virgil';
                        src: url("${virgilFontOnDisk}");
                        font-display: swap;
                    }
            
                    /* https://github.com/microsoft/cascadia-code */
                    @font-face {
                        font-family: 'Cascadia';
                        src: url("${cascadiaFontOnDisk}");
                        font-display: swap;
                    }
                </style>
            
                <link href="${virgilFontOnDisk}" as="font" type="font/woff2" crossorigin="anonymous" />
                <link href="${cascadiaFontOnDisk}" as="font" type="font/woff2" crossorigin="anonymous" />
            
                <link rel="stylesheet" type="text/css" href="${styleScriptOnDisk}" />
                <style>
                    .LoadingMessage {
                        position: fixed;
                        top: 0;
                        right: 0;
                        bottom: 0;
                        left: 0;
                        z-index: 999;
                        display: flex;
                        align-items: center;
                        justify-content: center;
                        pointer-events: none;
                    }
            
                    .LoadingMessage span {
                        background-color: rgba(255, 255, 255, 0.8);
                        border-radius: 5px;
                        padding: 0.8em 1.2em;
                        font-size: 1.3em;
                    }
            
                    .visually-hidden {
                        position: absolute !important;
                        height: 1px;
                        width: 1px;
                        overflow: hidden;
                        clip: rect(1px 1px 1px 1px);
                        clip: rect(1px, 1px, 1px, 1px);
                        white-space: nowrap;
                    }
            
                    .Stack_horizontal {
                        justify-content: start !important;
                    }
            

                    .App-menu_top>.Stack .Stack_horizontal>.ToolIcon_type_button:nth-child(2),
                    .App-menu_top>.Stack .Stack_horizontal>.ToolIcon_type_button:nth-child(5),
                    .App-menu_top>.Stack .Stack_horizontal>.ToolIcon_type_button:nth-child(6) {
                        display: none;
                    }
                </style>
            </head>
            
            <body>
                <div id="root">
                    <div class="LoadingMessage"><span>Loading scene...</span></div>
                </div>
            

                <script>
                    (() => {
                        let initData = {};
            
                        // https://github.com/Microsoft/vscode/issues/48464
                        Object.defineProperty(document, 'cookie', { value: '' });
            
                        // message from vscode
                        window.addEventListener('message', event => {
                            const message = event.data; // The JSON data our extension sent
            
                            const data = message.data;
            
                            console.log('message ===>', message);
            
                            switch (message.command) {
                                case 'loadLocalData':
                                    try {
                                        initData = JSON.parse(message.data || '{}');
                                        initLocalStorage();
                                        reloadScript();
                                    } catch (error) {
                                        throw new Error(error);
                                    }
            
                                    break;
                            }
                        });
            
                        function initLocalStorage() {
                            const vscode = acquireVsCodeApi();
            
                            const storage = {};
                            const log = console.log;
                            const bridgedLocalStorage = {
                                getItem: function (key) {
                                    log('localStorage: get ' + key);
            
                                    let result;
            
                                    switch (key) {
                                        case 'excalidraw':
                                            result = initData.elements;
                                            break;
                                        case 'excalidraw-state':
                                            result = initData.appState;
                                            break;
                                        default:
                                            result = storage[key];
                                            break;
                                    }

                                    return JSON.stringify(result);
                                },
                                setItem: function (key, val) {
                                    log('localStorage: set ' + key + ' to ' + val);
            
                                    switch (key) {
                                        case 'excalidraw':
                                            vscode.postMessage({
                                                command: 'updateExcalidraw',
                                                data: val
                                            });
                                            break;
            
                                        case 'excalidraw-state':
                                            vscode.postMessage({
                                                command: 'updateExcalidrawState',
                                                data: val
                                            });
                                    }
            
                                    storage[key] = val;
                                },
                                removeItem: function (key) {
                                    log('localStorage: remove ' + key);
                                    delete storage[key];
                                }
                            };
            
                            Object.defineProperty(window, 'localStorage', {
                                value: bridgedLocalStorage
                            });
                        }
            
                        function reloadScript() {
                            const fragment = document.createDocumentFragment();
                            const runtimeTag = document.createElement('script');
                            const mainScriptTag = document.createElement('script');
                            runtimeTag.src = '${runtimeScriptOnDisk}';
                            mainScriptTag.src = '${mainScriptOnDisk}';

                            fragment.appendChild(runtimeTag);
                            fragment.appendChild(mainScriptTag);
                            document.body.appendChild(fragment);
                        };

                        // rewrite command/ctrl + s
                        const isDarwin = /Mac|iPod|iPhone|iPad/.test(window.navigator.platform);
                        const CTRL_OR_CMD = isDarwin ? 'metaKey' : 'ctrlKey';
            
                        function patchFn(originalFn) {
                            return function (...args) {
                                const [eventName, oldHandler] = args;
                                console.log("Intercepting addListener", args);
                                if (eventName === 'keydown') {
                                    args[1] = e => {
                                        if (e.key === "s" && e[CTRL_OR_CMD]) {
                                            console.log('command/ctrl + s');
                                            return false;
                                        }
                                        return oldHandler(e);
                                    }
                                }
                                return originalFn.apply(this, args);
                            }
                        }
            
                        document.addEventListener = patchFn(document.addEventListener);
                    })();
                </script>

                <script src="${runtimeScriptOnDisk}"></script>
                ${chunkScript.map((item) => `<script src="${item}"></script>`)}
                <script src="${mainScriptOnDisk}"></script>
            </body>
            
            </html>
        `;
    }

    public resolveCustomTextEditor(document: vscode.TextDocument, webviewPanel: vscode.WebviewPanel) {
        webviewPanel.webview.options = {
            enableScripts: true,
            localResourceRoots: [vscode.Uri.file(join(this.buildPath))],
        };

        webviewPanel.webview.html = this.createWebViewContent(webviewPanel.webview);

        this.postFileContentToWebView(document, webviewPanel);

        webviewPanel.onDidDispose(() => {
            // ..
        });

        webviewPanel.webview.onDidReceiveMessage((e) => {
            switch (e.command) {
                case 'updateExcalidraw':
                    this.updateExcalidrawContent(document, e.data, 'updateExcalidraw');
                    return;

                case 'updateExcalidrawState':
                    this.updateExcalidrawContent(document, e.data, 'updateExcalidrawState');
            }
        });
    }

    public async postFileContentToWebView(document: vscode.TextDocument, webviewPanel: vscode.WebviewPanel) {
        const fileContent = await vscode.workspace.fs.readFile(document.uri);

        if (fileContent.length !== 0) {
            // init data to extension
            webviewPanel.webview.postMessage({ command: 'loadLocalData', data: fileContent.toString() });
        } else {
            // empty data
            const result = JSON.stringify({
                type: 'excalidraw',
                elements: [],
                appState: {
                    viewBackgroundColor: this.config.get(EXCALIDRAW_DEFAULT_BACKGROUNDCOLOR),
                },
            });
            webviewPanel.webview.postMessage({ command: 'loadLocalData', data: result });
        }
    }
}
